pkg/seedadmission/pods.go

```diff
@@ -0,0 +1,151 @@
+	// Ignore all resources other than our expected ones
+	switch request.Resource {
+	case
+		metav1.GroupVersionResource{Group: corev1.SchemeGroupVersion.Group, Version: corev1.SchemeGroupVersion.Version, Resource: "pods"}:
+	default:
+		return nil, nil
+	}
+
+	obj, err := getRequestObject(ctx, c, request)
+	if err != nil {
+		return nil, client.IgnoreNotFound(err)
+	}
+
+	entryLogger := gardenlogger.
+		NewFieldLogger(logger, "resource", fmt.Sprintf("%s/%s/%s", request.Kind.Group, request.Kind.Version, request.Kind.Kind)).
+		WithField("operation", request.Operation).
+		WithField("namespace", request.Namespace)
+
+	entryLogger.Info("Handling request")
+	patch, err := admitMutatePodAndMutate(ctx, c, entryLogger, obj, request.Namespace)
+	if err != nil {
+		logger.Errorf("Error mutating pod %s/%s: %v", request.Name, request.Namespace, err)
+		return nil, err
+	}
+
+	return patch, nil
+}
+
+// admitMutatePod checks if the object mutation is confirmed and return the pathces needed. If the given object is a list of objects then it
+// performs the check for every single object.
+func admitMutatePodAndMutate(ctx context.Context, c client.Client, logger *logrus.Entry, obj runtime.Object, namespace string) ([]jsonpatch.JsonPatchOperation, error) {
+	if strings.HasSuffix(obj.GetObjectKind().GroupVersionKind().Kind, "List") {
+		var patches []jsonpatch.JsonPatchOperation
+		err := meta.EachListItem(obj, func(o runtime.Object) error {
+			patch, err := mutateShootControlPlanePodAnnotations(ctx, c, logger, o, namespace)
+			if err != nil {
+				return err
+			}
+			patches = append(patches, patch...)
+			return nil
+		})
+		return patches, err
+	}
+
+	return mutateShootControlPlanePodAnnotations(ctx, c, logger, obj, namespace)
+}
+
+func mutateShootControlPlanePodAnnotations(ctx context.Context, c client.Client, logger *logrus.Entry, pod runtime.Object, namespace string) ([]jsonpatch.JsonPatchOperation, error) {
+	var patch []jsonpatch.JsonPatchOperation
+	// determine whether to perform mutation
+	isMutationRequired, err := mutationRequired(ctx, c, namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	acc, err := meta.Accessor(pod)
+	if err != nil {
+		return nil, err
+	}
+
+	if !isMutationRequired {
+		logger.Infof("Skipping mutation for %s/%s due to policy check", acc.GetNamespace(), acc.GetName())
+		return nil, nil
+	}
+
+	originalAnnotations := acc.GetAnnotations()
+	annotationsToAdd := map[string]string{
+		"fluentbit.io/exclude": "true",
+	}
+
+	for key, value := range annotationsToAdd {
+		if originalAnnotations == nil || originalAnnotations[key] == "" {
+			patch = append(patch, jsonpatch.JsonPatchOperation{
+				Operation: "add",
+				Path:      "/metadata/annotations",
+				Value: map[string]string{
+					key: value,
+				},
+			})
+		} else {
+			patch = append(patch, jsonpatch.JsonPatchOperation{
+				Operation: "replace",
+				Path:      "/metadata/annotations/" + key,
+				Value:     value,
+			})
+		}
```

Comment:
Doesn't this overwrite the whole annotations if there is at least one existing annotation?
Shouldn't it rather be:

```suggestion
		if originalAnnotations == nil {
			patch = append(patch, jsonpatch.JsonPatchOperation{
				Operation: "add",
				Path:      "/metadata/annotations",
				Value: map[string]string{
					key: value,
				},
			})
		} else if originalAnnotations[key] == "" {
		  patch = append(patch, jsonpatch.JsonPatchOperation{
				Operation: "add",
				Path:      "/metadata/annotations/" + key,
				Value: map[string]string{
					key: value,
				},
			})
		} else {
			patch = append(patch, jsonpatch.JsonPatchOperation{
				Operation: "replace",
				Path:      "/metadata/annotations/" + key,
				Value:     value,
			})
		}
```

Reply:
It doesn't overwrite the existing annotations. 
According to the documentation [link](http://jsonpatch.com/#add) `add` operation insert an object into an array. During the test all of the annotation persisted after mutation.

Reply:
OK, sure, makes sense. Thank you!